// frontity/dist/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@emotion/core");
exports.css = core_1.css;
exports.Global = core_1.Global;
exports.keyframes = core_1.keyframes;
var styled_1 = require("@emotion/styled");
exports.styled = styled_1.default;
var react_helmet_async_1 = require("react-helmet-async");
exports.Head = react_helmet_async_1.Helmet;
exports.HelmetProvider = react_helmet_async_1.HelmetProvider;
var component_1 = require("@loadable/component");
exports.loadable = component_1.default;
var connect_1 = require("@frontity/connect");
exports.observe = connect_1.observe;
exports.batch = connect_1.batch;
const connect_2 = __importStar(require("@frontity/connect"));
exports.connect = connect_2.default;
exports.createStore = connect_2.createStore;
const node_fetch_1 = __importDefault(require("node-fetch"));
exports.fetch = node_fetch_1.default;
var url_1 = require("url");
exports.URL = url_1.URL;
